text/0000-throw-expr.md

```diff
@@ -0,0 +1,875 @@
+we get `throw { 42 }`. This is still fine and unambiguous.
+
+Let's now consider the expression `{}` which is typed at `()`,
+which we see in this example:
+```rust
+let x: () = {};
+```
+
+If we instead write `throw {}`, that is equivalent to `throw ()`,
+then this conflicts with the struct literal `throw {}`.
+Therefore, while the probability of `throw {}` being written is low,
+we still have an ambiguity and the keyword can't be contextual.
+
+## The keyword shouldn't be contextual
+
+[keyword policy]: https://paper.dropbox.com/doc/Keyword-policy-SmIMziXBzoQOEQmRgjJPm
+[permalink]: https://gist.github.com/Centril/4c82c19b3cb02cc565622a37d1591785
+
+A recent [keyword policy] ([permalink]), adopted by the language team,
+decided that moving forward, keywords for new features in new editions
+should be real keywords instead of being contextual. The main motivation
+for this was to optimize for maintenance (and reduce technical debt).
+
+## Choice of keyword
+[choice of keyword]: #choice-of-keyword
+
+We consider the following keywords:
+
++ `throw`
++ `raise`
++ `fail`
++ `die`
++ `error`
+
+We dismiss `die` because we don't want to refer to killing the process.
+
+A reason to reject `error` specifically is that `std::error` exists as a module,
+`std::io::IntoInnerError::error` is also a function that would be broken by
+introducing `error` as a keyword. This keyword is also the only one that is
+used in the standard library as an identifier.
+
+It could be considered a benefit of `fail` that it is close to the `failure`
+crate's `bail!` macro. However; the RFC author considers this to mainly be
+a drawback because it is easy to mistake one for the other.
+Meanwhile `error` and `fail` are too tied to failure and error handling.
+
+Since the words `die`, `error`, and `fail` are much less commonly used than
+the keywords `throw` and `raise` we move on to consider the latter two.
+
+The main benefit of `throw` is that while it is familiar in languages with
+exception handling, it is less indicative of exceptions. It makes sense to say
+*"throw a ball"* and it works with the "re-throw" framing of the `?` operator.
+
+In the RFC authors view, `raise` + a noun is used less often.
+However, one benefit of `raise` is that while you can `throw` something
+vertically and horizontally, you usually only `raise` something upwards.
+This is good because if we consider nested `try { .. }` expressions or scopes,
+then `raise` moves "up".
+
+Considering crates named as the keyword we find that there isn't a crate named
+`raise`, there is a crate named [`throw`](https://crates.io/crates/throw).
+This crate would not be directly usable on the new edition,
+but it has zero reverse dependencies. Therefore the extent of the breakage
+is deemed quite low.
+Searching for `raise` with sourcegraph indicates a very
+small number of uses as an identifier, but searching for `throw` gives a timeout
+indicating that there are no uses.
```

Comment:
I've got at least 4 results from https://sourcegraph.com/search?q=repogroup:crates+%5Cbfn%5Cs%2Bthrow%5Cb. 

All of them are JS bindings (e.g. [servo/servo](https://github.com/servo/servo/blob/245dcc2118c88f7b51d65c124a85d74dd356a77a/components/script/dom/bindings/interface.rs#L91) and [neon-bindings/neon](https://github.com/neon-bindings/neon/blob/d96c418ba86ff72c1760e472e15fe8438061b5ec/src/js/error.rs#L14)), but I guess it is because of sourcegraph's selection of repositories.

Reply:
Strange. I tried the following query before: https://sourcegraph.com/search?q=repogroup:crates+case:yes++\b((let|const|type|)\s%2Bthrow\s%2B%3D|(fn|impl|mod|struct|enum|union|trait)\s%2Bthrow)\b+max:400 and it timed out, but now it shows the results you reference.

Anyways, the breakage slightly is less extensive than for `raise` and `fail`; but all keywords in this space break minimally.

I'll update the RFC with this new info :)